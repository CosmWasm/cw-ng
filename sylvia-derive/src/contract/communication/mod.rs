pub mod api;
pub mod enum_msg;
pub mod executor;
pub mod instantiate_builder;
pub mod querier;
pub mod reply;
pub mod struct_msg;
pub mod wrapper_msg;
