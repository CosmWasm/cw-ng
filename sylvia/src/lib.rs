#![cfg_attr(docsrs, feature(doc_cfg))]
//! Framework for creating CosmWasm Smart Contract with high-level abstraction layer
//!
//! Most of implementation lies in `sylvia-derive` crate which is reexported here

pub mod builder;
pub mod into_response;
#[cfg_attr(docsrs, doc(cfg(feature = "mt")))]
#[cfg(feature = "mt")]
pub mod multitest;
pub mod types;
pub mod utils;

#[cfg_attr(docsrs, doc(cfg(feature = "mt")))]
#[cfg(feature = "mt")]
pub use anyhow;
#[cfg_attr(docsrs, doc(cfg(feature = "mt")))]
#[cfg(feature = "mt")]
pub use cw_multi_test;
#[cfg_attr(docsrs, doc(cfg(feature = "cosmwasm_1_2")))]
#[cfg(feature = "cosmwasm_1_2")]
pub use cw_utils;
#[cfg_attr(docsrs, doc(cfg(feature = "mt")))]
#[cfg(feature = "mt")]
pub use derivative;
pub use sylvia_derive::{contract, entry_points, interface};
pub use {
    cosmwasm_schema as cw_schema, cosmwasm_std as cw_std, schemars, serde,
    serde_cw_value as serde_value, serde_json_wasm as serde_json,
};

pub mod replies {
    use cosmwasm_std::{DepsMut, Empty, Env, Event, MsgResponse};

    /// Represantation of `reply` context received in entry point.
    #[deprecated(
        since = "1.3.0",
        note = "This type is added temporarily to not break existing API. Since `2.0.0` it will replace the `sylvia::types::ReplyCtx` type."
    )]
    #[non_exhaustive]
    pub struct ReplyCtx<'a, C: cosmwasm_std::CustomQuery = Empty> {
        pub deps: DepsMut<'a, C>,
        pub env: Env,
        pub gas_used: u64,
        pub events: Vec<Event>,
        pub msg_responses: Vec<MsgResponse>,
    }

    #[allow(deprecated)]
    impl<'a, C: cosmwasm_std::CustomQuery>
        From<(DepsMut<'a, C>, Env, u64, Vec<Event>, Vec<MsgResponse>)> for ReplyCtx<'a, C>
    {
        fn from(
            (deps, env, gas_used, events, msg_responses): (
                DepsMut<'a, C>,
                Env,
                u64,
                Vec<Event>,
                Vec<MsgResponse>,
            ),
        ) -> Self {
            Self {
                deps,
                env,
                gas_used,
                events,
                msg_responses,
            }
        }
    }
}
