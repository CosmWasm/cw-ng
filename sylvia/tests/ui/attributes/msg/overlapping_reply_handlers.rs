#![allow(unused_imports)]
#![cfg(feature = "sv_replies")]

use sylvia::contract;
use sylvia::cw_std::{Reply, Response, StdResult};
use sylvia::types::{InstantiateCtx, ReplyCtx};

pub struct Contract;

#[contract]
impl Contract {
    pub fn new() -> Self {
        Self
    }

    #[sv::msg(instantiate)]
    pub fn instantiate(&self, _ctx: InstantiateCtx) -> StdResult<Response> {
        Ok(Response::new())
    }

    #[sv::msg(reply, handlers(handler1))]
    fn reply_always(&self, _ctx: ReplyCtx, _reply: Reply) -> StdResult<Response> {
        Ok(Response::new())
    }

    #[sv::msg(reply, handlers(handler1), reply_on=success)]
    fn duplicated_success_for_reply_always(
        &self,
        _ctx: ReplyCtx,
        _reply: Reply,
    ) -> StdResult<Response> {
        Ok(Response::new())
    }

    #[sv::msg(reply, handlers(handler2), reply_on=failure)]
    fn some_reply(&self, _ctx: ReplyCtx, _reply: Reply) -> StdResult<Response> {
        Ok(Response::new())
    }

    #[sv::msg(reply, reply_on=failure)]
    fn handler2(&self, _ctx: ReplyCtx, _reply: Reply) -> StdResult<Response> {
        Ok(Response::new())
    }
}

fn main() {}
