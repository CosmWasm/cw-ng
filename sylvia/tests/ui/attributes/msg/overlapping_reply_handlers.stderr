error: Duplicated reply handler.

         = note: Previous definition of handler=HANDLER_1_REPLY_ID for reply_on=always defined on `fn reply_always()`

  --> tests/ui/attributes/msg/overlapping_reply_handlers.rs:26:32
   |
26 |     #[sv::msg(reply, handlers=[handler1], reply_on=success)]
   |                                ^^^^^^^^

error: Duplicated reply handler.

         = note: Previous definition of handler=HANDLER_2_REPLY_ID for reply_on=failure defined on `fn some_reply()`

  --> tests/ui/attributes/msg/overlapping_reply_handlers.rs:41:8
   |
41 |     fn handler2(&self, _ctx: ReplyCtx, _reply: Reply) -> StdResult<Response> {
   |        ^^^^^^^^
